### DISCUSSION TITLE
形码不能生成词组编码的问题〔包含输入全码之前让候选按词频排序的办法〕

### DISCUSSION THREAD

**Discussion by Marksusu** [rime/home#1409] [General]
我遇到的问题是──词组不能根据单字生成编码。

在 librime 下面留言后才发现这里也能做讨论，正好我也不确定问题是否就出在 librime ，就展开说说我所依赖的功能和我目前遇到的问题吧。

在使用形码时，有个“生成词组编码”的功能，可以按照字数设定规则，根据单字的编码，算出相应的词组的编码。
这个功能非常好用，结合 table_translator ，可以在一个主方案里，引入多重编码规则。
这样做的好处是，无论输入一码、两码、三码还是四码，词组都能按照词频排序。

通常，形码输入都是要在输入四码全码的时候，列表里的词组才会按照词频排序，不足四码时，都是按照未补全的字母排序的。

〔因为我目前的部署是失灵的，所以没法制作如上描述的截图，但用过形码的人中间一定有人会注意到如上描述的现象。〕

既然只有四码全码才能参与排序，那就制做三码和两码为全码的辅助方案，再放入四码为全码的主方案中，就能实现如下的简词效果：

![截图_2023-11-14_11-12-52](https://github.com/rime/home/assets/109450118/1e8ac33a-dbce-49b7-a95d-3362801f8870)

【**▲图一**：当输入 `isn` 三码时，三字词组都是按照词频排序的──这就是“三码全码”的辅助方案在起作用。】
【值得注意的是：图中，第 9 、第 0 候选是“双字词”，不在“三码全码”的规则之内，所以当输入 `isn` 时，没有相应的“全码候选”，所以，这两个候选的排序就不“词频序”而是“字母序”。】

![截图_2023-11-14_11-10-35](https://github.com/rime/home/assets/109450118/d2b3b53c-c690-41e2-a99d-55e51b33fe43)

【**▲图二**：当输入 `is` 两码时，“两码全码”的辅助方案起了作用，实现了“两码简词”的自动排序。】

只要把“有编码的单字”和“无编码的词组”分别放在不同的词库文件里，再分别于不同的方案中指定相应字数词组的相应编码规则，再用 table_translator 合成在一个主方案中，就能让每次击键的结果参与排序。

这样做，就能让选重变得十分方便。
并且，你还能随时调整词库中的词组和词频，重新部署后就能更新词组的编码和排序，非常好用。
你还可以随时更改单字的编码，重新部署后，所有涉及该单字的词组编码都会相应地更新，实在方便极了。

但这些便利极度依赖“生成编码”的功能，这个功能失灵后，重新部署后就只能打单字了。

![截图_2023-11-14_11-49-28](https://github.com/rime/home/assets/109450118/2027a6aa-a496-456a-b592-a4b70f470d5e)

【**▲图三**：“词组生成编码”失灵，输入 `is` 时，仅有单字候选，“两码全码”的词组候选一个也没有出现。】

![截图_2023-11-14_11-51-21](https://github.com/rime/home/assets/109450118/2ffcc7d5-314a-4029-b2ba-be820b4f5ddf)

【**▲图四**：“词组生成编码”失灵，输入 `isn`，本该出现的三字词组──“三码全码”候选一个也没出现。】

我用的系统是 garuda linux ，安装的是 ibus-rime 1.5.0-1，还有 librime-1:1.9.0-1-x86_64 ，我不清楚是不是它的问题。
我现在只能用早前备份的 build 文件夹中的二进制文件去替换新生成的，才能回到之前能打词组的状态。

![截图_2023-11-16_11-46-11](https://github.com/rime/home/assets/109450118/09e2990c-0b62-4808-9071-4ad0a7d16a42)

【**▲图五**：build 文件夹中的情况。左边是新生成的文件，右边是我之前备份的文件。可以看到，“词组生成编码”失灵后，通过“部署”生成的二进制文件明显变小了。】

请各位大佬帮我解惑，谢谢～

**Comment by MokOopsing** [comment]
把你的dict文件头拿出来看看你的规则

**Reply by Marksusu** [reply]
https://github.com/Marksusu/Al2O3

我学会上传文件了，我还写了简单的说明。

---

昨天在小狼毫上试了一下，也是没问题的。

目前我能看到的区别就是：一个是 rime_version: 1.8.5 ；
另一个是 rime_version: 1.9.0 。
